### PR TITLE
feat: DBTP-1380/DBT-1393 validate Redis and OpenSearch versions are valid and supported versions

### PR DESCRIPTION
**Pull request by mramplin**
Addresses [DBTP-1380](https://uktrade.atlassian.net/browse/DBTP-1380) and [DBTP-1393](https://uktrade.atlassian.net/browse/DBTP-1393)

Please add any relevant context for you pull request here, or delete this if none needed.

---
### Title:
- [x] Scope included as per [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Ticket reference included (unless it's a quick out of ticket thing)
### Description:
- [x] Link to ticket included (unless it's a quick out of ticket thing)
- [x] Includes tests (or an explanation for why it doesn't)
- [x] If the work includes user interface changes, before and after screenshots included in description
- [x] Includes any applicable changes to the documentation in this code base
- [x] Includes link(s) to any applicable changes to the documentation in the [DBT Platform Documentation](https://platform.readme.trade.gov.uk/) (can be to a pull request)
### Tasks:
- [x] [Trigger the pull request regression tests for this branch](https://github.com/uktrade/platform-tools?tab=readme-ov-file#regression-tests) and confirm that they are passing
Regression Tests Completed [here](https://eu-west-2.console.aws.amazon.com/codesuite/codepipeline/pipelines/pull-request-regression-tests/executions/3ac0978e-34c5-43b0-8fb3-cd9316a7c9b7/visualization?region=eu-west-2)

[DBTP-1380]: https://uktrade.atlassian.net/browse/DBTP-1380?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DBTP-1393]: https://uktrade.atlassian.net/browse/DBTP-1393?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mramplin** [comment]
Docs error on lint checks seems to be a false positive as returns a successful message but a 1 exit code.

**Review by chiaramapellimt — APPROVED** [review]

File: dbt_platform_helper/utils/files.py [hunk 1]
```diff
@@ -102,3 +106,76 @@ def combine_env_data(data):
...
+
+    An API call is needed if any of the following conditions are met:
+        1. No cache file (.platform-helper-config.yml) exists.
+        2. The resource name (e.g. redis, opensearch) does not exist within the cache file.
+        3. The date-retrieved value of the cached data is > than a time interval. In this case 1 day.
+    """
+
+    api_call_required = True
+
+    if os.path.exists(platform_helper_config_file):
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/files.py:152]: Can turn this into a guard clause - 

> if !os.path.exists(...) return True

Allows us to exit early and minimise the levels of nesting required.

File: dbt_platform_helper/utils/files.py [hunk 2]
```diff
@@ -102,3 +106,76 @@ def combine_env_data(data):
...
+
+        platform_helper_config = read_file_as_yaml(platform_helper_config_file)
+
+        if platform_helper_config.get(resource_name):
+
+            api_call_required = check_if_cached_datetime_is_greater_than_interval(
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/files.py:158]: By using guard clauses and reducing the nesting, we wouldn't need to store the result at all, and could instead return it directly

File: dbt_platform_helper/utils/validation.py [hunk 3]
```diff
@@ -553,8 +559,91 @@ def validate_platform_config(config, disable_aws_validation=False):
...
+    for redis_extension in redis_extensions:
+        environments = redis_extension.get("environments", {})
+        if isinstance(environments, dict):
+            for environment in environments.keys():
+                env_config = environments.get(environment, {})
+                if isinstance(env_config, dict):
+                    redis_engine_version = env_config.get("engine")
+                    if redis_engine_version not in supported_redis_versions:
+                        redis_environments_with_failure.append(
+                            [extension, environment, redis_engine_version]
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/validation.py:595]: This is 5 levels of nesting. Could be useful to break into methods to make it clear what each level is trying to do, and help turn some of the if statements into guard clauses.

File: dbt_platform_helper/utils/validation.py [hunk 4]
```diff
@@ -553,8 +559,91 @@ def validate_platform_config(config, disable_aws_validation=False):
...
+    if cache_refresh_required("redis"):
+        supported_redis_versions = get_redis_supported_versions()
+    else:
+        supported_redis_versions = read_supported_versions_from_cache("redis")
+
+    redis_extensions = []
+    redis_environments_with_failure = []
+
+    for extension in config.get("extensions").keys():
+
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/validation.py:582]: Random blank line?

File: dbt_platform_helper/utils/validation.py [hunk 5]
```diff
@@ -553,8 +559,91 @@ def validate_platform_config(config, disable_aws_validation=False):
...
+    if cache_refresh_required("opensearch"):
+        supported_opensearch_versions = get_opensearch_supported_versions()
+    else:
+        supported_opensearch_versions = read_supported_versions_from_cache("opensearch")
+
+    opensearch_extensions = []
+    opensearch_environments_with_failure = []
+
+    for extension in config.get("extensions", {}).keys():
+
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/validation.py:621]: Random blank line?

File: dbt_platform_helper/utils/validation.py [hunk 6]
```diff
@@ -553,8 +559,91 @@ def validate_platform_config(config, disable_aws_validation=False):
...
+        if config.get("extensions").get(extension).get("type") == "opensearch":
+            opensearch_extensions.append(config["extensions"][extension])
+
+    for opensearch_extension in opensearch_extensions:
+
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/validation.py:626]: Random blank line?

File: dbt_platform_helper/utils/validation.py [hunk 7]
```diff
@@ -553,8 +559,91 @@ def validate_platform_config(config, disable_aws_validation=False):
...
+        environments = opensearch_extension.get("environments", {})
+        if isinstance(environments, dict):
+            for environment in environments.keys():
+                env_config = environments.get(environment, {})
+                if isinstance(env_config, dict):
+                    opensearch_engine_version = env_config.get("engine")
+                    if opensearch_engine_version not in supported_opensearch_versions:
```
Comment (benn-dbt) [inline dbt_platform_helper/utils/validation.py:633]: This is 5 levels of nesting. Could be useful to break into methods to make it clear what each level is trying to do, and help turn some of the if statements into guard clauses.

File: dbt_platform_helper/COMMANDS.md [hunk 8]
```diff
@@ -40,8 +40,7 @@
 - [platform-helper notify environment-progress](#platform-helper-notify-environment-progress)
 - [platform-helper notify add-comment](#platform-helper-notify-add-comment)
 - [platform-helper database](#platform-helper-database)
-- [platform-helper database dump](#platform-helper-database-dump)
-- [platform-helper database load](#platform-helper-database-load)
+- [platform-helper database copy](#platform-helper-database-copy)
```
[the remark this answers is not in the document]
Comment (WillGibson) [reply dbt_platform_helper/COMMANDS.md:43]: This file is auto generated, so almost certainly the behind `main` thing.

File: dbt_platform_helper/utils/validation.py (hunk 7 above)
Comment (WillGibson) [reply dbt_platform_helper/utils/validation.py:633]: It looks very similar to the 5 deep nesting above too. There may be an opportunity to extract some or all of it out to shared method(s).

**Comment by DeveloperConnor** [comment]
The comments on guard clauses/too much indentation sound reasonable and will add them. Not sure I agree on the random blank lines as they are there for readability, and aren't a failure under PEP 8.

File: dbt_platform_helper/COMMANDS.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (DeveloperConnor) [reply dbt_platform_helper/COMMANDS.md]: Fixed 👍

File: dbt_platform_helper/utils/files.py (hunk 1 above)
Comment (DeveloperConnor) [reply dbt_platform_helper/utils/files.py:152]: That would just make the method longer? Also not a fan of if statements that use !/nots as it just makes reading the code 1 year later even more confusing.

    if not os.path.exists(platform_helper_config_file):
        return cache_refresh_required

    platform_helper_config = read_file_as_yaml(platform_helper_config_file)

    if platform_helper_config.get(resource_name):

        cache_refresh_required = check_if_cached_datetime_is_greater_than_interval(
            platform_helper_config[resource_name].get("date-retrieved"), 1
        )

    return cache_refresh_required

Comment (DeveloperConnor) [reply dbt_platform_helper/utils/files.py:152]: Regardless - pushed some changes to address 👍

File: dbt_platform_helper/utils/validation.py (hunk 7 above)
Comment (DeveloperConnor) [reply dbt_platform_helper/utils/validation.py:636]: Agreed  @WillGibson @benn-dbt it reads badly - will push some changes

File: dbt_platform_helper/COMMANDS.md
Comment (WillGibson) [inline dbt_platform_helper/COMMANDS.md:1]: Some conflicts in here. Should be resolved by regenerating this file after merging in `main`.

File: dbt_platform_helper/utils/files.py [hunk 9]
```diff
@@ -7,6 +9,8 @@
 from jinja2 import Environment
 from jinja2 import FileSystemLoader
 
+platform_helper_config_file = ".platform-helper-config.yml"
```
Comment (WillGibson) [inline dbt_platform_helper/utils/files.py:12]: Is this referring to what we would normally call the `platform_config_file`?

Comment (DeveloperConnor) [reply dbt_platform_helper/utils/files.py:12]: No - its the cache file.

Comment (WillGibson) [reply dbt_platform_helper/utils/files.py:12]: Can we rename it to make that more obvious.

File: dbt_platform_helper/COMMANDS.md
Comment (DeveloperConnor) [reply dbt_platform_helper/COMMANDS.md:1]: Good spot - sorted it.

**Pull request merged**